src/include/homestore/blk.h

```diff
@@ -24,122 +24,162 @@
 #include <memory>
 #include <sstream>
 
+#include <boost/icl/interval_map.hpp>
 #include <sisl/utility/enum.hpp>
+#include <sisl/fds/buffer.hpp>
 #include <homestore/homestore_decl.hpp>
 
 namespace homestore {
 
-typedef uint32_t blk_num_t;
-typedef blk_num_t blk_cap_t;
-static_assert(sizeof(blk_num_t) == (BLK_NUM_BITS - 1) / 8 + 1, "Expected blk_num_t to matching BLK_NUM_BITS");
+using chunk_num_t = uint16_t;
+using blk_count_t = uint16_t;
+using blk_num_t = uint32_t;
+using blk_temp_t = uint16_t;
 
-typedef uint8_t blk_count_serialized_t;
-typedef uint16_t blk_count_t;
-static_assert(sizeof(blk_count_serialized_t) == (NBLKS_BITS - 1) / 8 + 1,
-              "Expected blk_count_t to matching NBLKS_BITS");
+static constexpr size_t max_addressable_chunks() { return 1UL << (8 * sizeof(chunk_num_t)); }
+static constexpr size_t max_blks_per_chunk() { return 1UL << (8 * sizeof(blk_num_t)); }
+static constexpr size_t max_blks_per_blkid() { return (1UL << (8 * sizeof(blk_count_t))) - 1; }
 
-typedef uint8_t chunk_num_t;
-static_assert(sizeof(chunk_num_t) == (CHUNK_NUM_BITS - 1) / 8 + 1, "Expected blk_count_t to matching CHUNK_NUM_BITS");
-
-typedef uint8_t blk_temp_t;
-
-/* This structure represents the application wide unique block number. It also encomposses the number of blks. */
+#pragma pack(1)
 struct BlkId {
-private:
-    static constexpr uint64_t s_blk_num_mask{(static_cast< uint64_t >(1) << BLK_NUM_BITS) - 1};
-    static constexpr uint64_t s_nblks_mask{(static_cast< uint64_t >(1) << NBLKS_BITS) - 1};
-    static constexpr uint64_t s_chunk_num_mask{(static_cast< uint64_t >(1) << CHUNK_NUM_BITS) - 1};
+protected:
+    struct serialized {
+        blk_num_t m_is_multi : 1; // Is it a part of multi blkid or not
```

Comment:
I probably didn't get the reason BlkId has to understand this `m_is_multi`? 
What will be the problem if BlkId doesn't know about `m_is_multi`, I think MultiBlkId should just work as it is, no?

Reply:
@xiaoxichen, we don't rely on `m_is_multi` to know how to deserialize the BlkID/MultiBlkId. It is a compare of blob.size vs sizeof(BlkId) to know whether this is a MultiBlkId of a BlkId. 

I think the reason is a lot of places is using `bid.is_multi()` to determine the need to access multi blkids. 
But it also be achieved by `dynamic_cast<MultiBlkId*>` with BlkId knows nothing about how derived class is using this base class. I could missing some other use cases that forbid us from doing above and would like to learn what are they.

Reply:
> bool type might be more readable

bool type size is not guaranteed across and it uses min 1 byte, whereas we need to serialize as 1 bit.

Reply:
The MultiBlkId is not a virtual derived class of BlkId. It is present right now as a way for upper layer to continue to use BlkId in-place of MultiBlkId. For example, metablk uses the BlkId everywhere and I wanted to make sure their usage is not broken. Once we address in all the places, we can make MultiBlkId as non-derived class of BlkId.

Reply:
I understand it require a lot of dyamic_cast<BlkID> to be added for a virtual derived class. 
It seems a virtual derived class makes more sense, with BlkId takens care of single blkid, and MultiBlkId takens care of fragmented BlkId, and single blkid's business be taken care by its base class. 

Or non-derived version with BlkId as a container of fragmented blkids.

Totally fine if it needs to be done in two phases, if it is two much in one shot. You can decide how to approach.

Reply:
We can't do vtable if we are going to directly serialize this, since sizeof(BlkId) will become 16 bytes instead of 8 bytes. I don't see much value in making BlkId being virtual in any case. In fact we should try to reverse, where I am thinking of making MultiBlkId independent of BlkId. Only reason it is done today is for backward compatibility (especially metablks directly persists BlkId) in persistent structures.

Reply:
I am fine doing non-derived version.